protocol/transport.go

```diff
@@ -1,68 +1,28 @@
 package protocol
 
 import (
-	"encoding/binary"
-	"fmt"
+	"github.com/jackc/pgx/pgproto3"
 	"io"
 )
 
-// NewTransport creates a protocol
-func NewTransport(r io.Reader, w io.Writer) *Transport {
+// NewTransport creates a Transport
+func NewTransport(rw io.ReadWriter) *Transport {
+	b, _ := pgproto3.NewBackend(rw, nil)
 	return &Transport{
-		R: r,
-		W: w,
+		w: rw,
+		r: b,
 	}
 }
 
 // Transport manages the underlying wire protocol between backend and frontend.
 type Transport struct {
-	R           io.Reader
-	W           io.Writer
-	initialized bool
+	w           io.Writer
+	r           *pgproto3.Backend
 	transaction *transaction
 }
 
-// StartUp handles the very first messages exchange between frontend and backend of new session
-func (t *Transport) StartUp() (Message, error) {
-	// read the initial connection startup message
-	msg, err := t.Read()
-	if err != nil {
-		return nil, err
-	}
-
-	if msg.IsCancel() {
-		return msg, nil
-	}
-
-	if msg.IsTLSRequest() {
-		// currently we don't support TLS.
-		err := t.Write(TLSResponse(false))
-		if err != nil {
-			return nil, err
-		}
-
-		msg, err = t.Read()
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	v, err := msg.StartupVersion()
-	if err != nil {
-		return nil, err
-	}
-
-	if v != "3.0" {
-		return nil, fmt.Errorf("unsupported protocol version %s", v)
-	}
-
-	t.initialized = true
-
-	return msg, nil
-}
-
 func (t *Transport) beginTransaction() {
-	t.transaction = &transaction{transport: t, in: []Message{}, out: []Message{}}
+	t.transaction = &transaction{transport: t, in: []pgproto3.FrontendMessage{}, out: []Message{}}
```

Comment:
this will have the same effect :)
```suggestion
	t.transaction = &transaction{transport: t}
```

(applies also to tests, but less important)